karapace/master_coordinator.py

```diff
@@ -49,16 +50,25 @@ def group_protocols(self):
 
     def _perform_assignment(self, leader_id, protocol, members):
         self.log.info("Creating assignment: %r, protocol: %r, members: %r", leader_id, protocol, members)
-        self.master = None
+        self.are_we_master = None
         error = NO_ERROR
         urls = {}
+        fallback_urls = {}
         for member_id, member_data in members:
             member_identity = json.loads(member_data.decode("utf8"))
             if member_identity["master_eligibility"] is True:
                 urls[get_identity_url(member_identity["scheme"], member_identity["host"],
                                       member_identity["port"])] = (member_id, member_data)
-        self.master_url = sorted(urls, reverse=self.election_strategy.lower() == "highest")[0]
```

Comment:
I didn't get your point here. This line is old code, and it is the problematic accessing of the array out of bounds. In this PR there is a fix that the array is not accessed out-of-bounds.

The ordering has been by DNS/IP, and AFAIK it works. The values of `member_id` are originating Kafka communication.

Reply:
> I didn't get your point here. This line is old code, and it is the problematic accessing of the array out of bounds. In this PR there is a fix that the array is not accessed out-of-bounds.

The point I'm raising here is not about the `IndexError`, it is about the `sorted` itself. I know the code was not introduced by this PR, but when the code was introduced really doesn't change if it works as intended or not.

Regarding the `sort` itself. I don't think there is any order in IP addresses or DNS names. This leaves two options:

1. Kafka shuffles the data of `members` around for some reason, and the original `sort` was not about finding the newest node, but about having a stable result on subsequent runs (otherwise the master would change on every run).
2. The original sort was added to order nodes by age, but unfortunately used inadequate data.

Regardless of which one happened, the second commit tried to improve the code based on the sort order, but it doesn't work as expected. So, I raised the point with `member_id` as a candidate to fix the issue. Instead of sorting by IP/DNS, maybe sorting by `member_id` would work. I just don't know the semantics of the value by heart and it needs to be checked if it will fix the issue.